<a href="https://colab.research.google.com/github/manfredialdo/iad-tpi01/blob/main/Manfredi_Aldo_Comision10_Entrega2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## . Datos de identificación

- **Apellido y nombre:** manfredi aldo dario
- **Comisión:** 10
- **Dataset asignado:** hotel booking TPI grupo J.csv
- **Entrega:** Trabajo practico integrador primera entrega
- **Fecha:** Domingo 6 de septiembre de 2026, 23:59 hs.
- **Repositorio:** https://github.com/manfredialdo/iad-tpi01
- **Archivo:** Manfredi_Aldo_Comision10_Entrega2.ipynb

# 🧩 TPI · Entrega 2 · Limpieza, preparación y transformación de datos (OBLIGATORIA)

En esta entrega se debe aplicar al dataset asignado los procedimientos trabajados durante la Unidad N° 2. El objetivo es diagnosticar la calidad de los datos, justificar decisiones de limpieza y construir las transformaciones necesarias para preparar el dataset para las próximas etapas del análisis.

* **Caso de estudio:** Reservas hoteleras
* **Variable objetivo:** `is_canceled`
* **Lenguaje a elección:** Python o R (debes elegir y utilizar un solo lenguaje).
* **⏱️ Tiempo estimado:** 2 pomodoros de 20 minutos (continuidad de la Semana 4).

---

## 💻 Datos del Estudiante e Identificación

> **Atención:** Completa tus datos personales antes de comenzar con la resolución del notebook.

* **Apellido y Nombre:** manfredi aldo dario
* **Comisión:** 10
* **Número de Entrega:** Entrega 2
* **Dataset Asignado:**  dataset J https://tup.sied.utn.edu.ar/pluginfile.php/32860/course/section/4282/hotel%20booking%20TPI%20grupo%20J.csv
* **Problema de Análisis:**
* **Variable Objetivo:** `is_canceled`

---

## 📝 Requisitos Técnicos de Carga

1. **Notebook independiente:** Debe poder ejecutarse desde el inicio de manera fluida sin depender de archivos de entregas anteriores.
2. **Carga reproducible:** Importación de bibliotecas necesarias y carga directa del dataset original.
3. **Copia de trabajo:** Conservar el dataset original intacto en una variable y crear una copia explícita sobre la cual se aplicarán las modificaciones.

---

## 📋 Contenido Obligatorio de la Entrega

1. **Diagnóstico de calidad:** Revisión de valores faltantes, duplicados, tipos de datos, categorías inconsistentes, valores inválidos y posibles valores atípicos, manteniendo la bitácora correspondiente.
2. **Transformación de variables:** Aplicación justificada de conversiones temporales, codificación, discretización, normalización o estandarización cuando resulte pertinente.
3. **Variables derivadas:** Creación de nuevas columnas útiles para el problema (ej. duración total de la estadía, cantidad total de huéspedes, importe estimado, reserva familiar o segmentos de anticipación).
4. **Bitácora del proceso:** Registro estructurado con las variables transformadas, el problema detectado, la decisión aplicada y su justificación.
5. **Dataset preparado:** Exposición del estado final del conjunto de datos y explicación de cómo las decisiones tomadas lo dejan listo para análisis posteriores.

> 📌 **Criterio metodológico:** No se espera que apliques todas las transformaciones posibles ni que elimines automáticamente cada valor inusual. Cada procedimiento debe responder a un problema o necesidad concreta y estar acompañado por una justificación.

---

## 🔎 Variables Sugeridas para Revisión

| Variable(s) | Aspecto a revisar / Transformación |
| :--- | :--- |
| `children`, `adults`, `babies` | Cantidad total de huéspedes e Inconsistencias. |
| `stays_in_weekend_nights`, `stays_in_week_nights` | Duración total de la estadía. |
| `adr` | Valores faltantes, negativos, ceros o atípicos. |
| `lead_time` | Anticipación de la reserva y valores elevados. |
| `country`, `agent`, `company` | Presencia e interpretación de valores ausentes. |
| `hotel`, `market_segment`, `distribution_channel`, `deposit_type`, `customer_type` | Revisión y posible codificación de variables categóricas. |
| **Variables de fecha de llegada** | Formato de fecha y construcción de variables temporales. |

---

## ✅ Lista de Comprobación (Checklist de Entrega)

- [ ] ¿El notebook identifica al estudiante, la comisión y la entrega?
- [ ] ¿Puede ejecutarse de principio a fin sin depender de otro notebook?
- [ ] ¿Conserva el dataset original intacto y aplica los cambios sobre una copia?
- [ ] ¿Revisa los principales problemas de calidad de los datos?
- [ ] ¿Justifica formalmente cada decisión de limpieza y transformación?
- [ ] ¿Las variables derivadas están correctamente construidas y calculadas?
- [ ] ¿Combina explicaciones en texto con bloques de código y sus resultados?
- [ ] ¿Explica detalladamente cómo quedó preparado el dataset final?

---

## ⚠️ Condiciones y Pautas de Entrega

* **Estado:** Obligatoria
* **Habilitación:** Lunes 31 de agosto de 2026, 00:00 hs.
* **Fecha límite de entrega:** Domingo 6 de septiembre de 2026, 23:59 hs.
* **Formato:** Único archivo `.ipynb`.
* **Tamaño máximo:** 20 MB.
* **Calificación mínima de aprobación:** 6 / 10 puntos.
* **Intentos permitidos:** Hasta 2 intentos (la reaparición será habilitada por el tutor cuando corresponda).
* **Estructura de nombre del archivo:** `Apellido_Nombre_Comision_Entrega2.ipynb`

> 💡 **Recomendación:** Conserva también una versión editable acumulativa (notebook maestro) donde integres progresivamente las distintas etapas del TPI para la entrega final.

1. **importacion de librerias, carga, copia explícita, plantilla bitacora**

In [5]:
import pandas as pd
import numpy as np

# 1. Carga reproducible y copia explícita
url = "https://raw.githubusercontent.com/manfredialdo/iad-tpi01/main/hotel%20booking%20TPI%20grupo%20J.csv"
df = pd.read_csv(url)
df_limpio = df.copy()

# Inicialización de Bitácora
registros_bitacora = []

def registrar(problema, variable, decision, justificacion):
    registros_bitacora.append({
        "problema_detectado": problema,
        "variable_afectada": variable,
        "decision_tomada": decision,
        "justificacion": justificacion
    })


In [ ]:
# Estructura de tipos de datos y conteo de nulos inicial
df_limpio.info()

# Resumen estadístico numérico
display(df_limpio.describe())

# Resumen estadístico categórico
display(df_limpio.describe(include=['object', 'category']))

**Conclusiones parciales del Análisis Exploratorio**

Cancelaciones y Anticipación: La tasa de cancelación es del 37%. Las reservas se hacen con mucha antelación (media de 104 días) y predomina la política sin depósito (87.5%), lo que facilita cancelaciones sin costo.  

Canales de Venta: City Hotel concentra el 66.5% de la demanda. El 81.5% de las reservas provienen de agencias/operadores (TA/TO) y el 40.8% de los clientes son de Portugal (PRT).  

Perfil del Huésped: Predominan los clientes individuales (75.2%) sin niños. La estadía promedio es de 3.4 noches y solo el 3.3% son huéspedes recurrentes.  

Tarifas y Habitaciones: El precio promedio (adr) es de $101.68. El plan de comida preferido es BB (Bed & Breakfast) y la habitación tipo A es la más solicitada y asignada.  

Calidad de Datos:

Faltantes: agent (13.9%) y company (94.2%) tienen nulos porque representan ventas directas o particulares. country registra solo 117 nulos (0.47%).  

Atípicos: Existen valores extremos como reservas con hasta 40 adultos, tarifas de $0.00 y esperas de 391 días.

# 2. Diagnóstico y Limpieza
## A. Duplicados Implícitos

In [6]:
# 2. Diagnóstico y Limpieza de Calidad
# la columna booking_id hace que cada fila parezca única, así que filtramos por
# las variables operativas reales para detectar si existían reservas idénticas creadas en simultáneo por agencias de viajes.
# Definimos las columnas clave de una reserva para identificar y contar cuántos
#  registros son exactamente iguales entre sí (duplicados de negocio).
# A. Duplicados Implícitos
subset_identificacion = [
    'hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'lead_time', 'adults', 'children', 'babies', 'adr', 'reserved_room_type'
]
duplicados = df_limpio.duplicated(subset=subset_identificacion, keep=False).sum()
registrar(
    f"Se detectaron {duplicados} registros idénticos por combinación clave de reserva.",
    "Atributos combinados de reserva",
    "Conservar registros registrando el hallazgo.",
    "Pueden responder a reservas simultáneas de grupos gestionadas en lote por la misma agencia."
)



In [8]:
duplicados

np.int64(7332)

existen 7332 registros que comparten exactamente los mismos datos de reserva (misma fecha, hotel, cantidad de huéspedes, tarifa y tipo de habitación)

In [9]:
registrar(
    f"Se detectaron {duplicados} registros idénticos por combinación clave de reserva.",
    "Atributos combinados de reserva",
    "Conservar registros registrando el hallazgo.",
    "Pueden responder a reservas simultáneas de grupos gestionadas en lote por la misma agencia."
)


# 2. Diagnóstico y Limpieza
## B. Reservas sin huéspedes

In [ ]:
# B. Reservas sin huéspedes
sin_huespedes = df_limpio[(df_limpio['adults'] + df_limpio['children'].fillna(0) + df_limpio['babies']) == 0]
if len(sin_huespedes) > 0:
    registrar(
        f"Existencia de {len(sin_huespedes)} reservas sin ningún huésped registrado.",
        "adults, children, babies",
        "Eliminar filas afectadas del dataset de trabajo.",
        "Una reserva requiere al menos 1 huésped registrado para ser comercialmente válida."
    )
    df_limpio = df_limpio.drop(sin_huespedes.index)


In [11]:
sin_huespedes = df_limpio[(df_limpio['adults'] + df_limpio['children'].fillna(0) + df_limpio['babies']) == 0]


In [12]:
sin_huespedes

,booking_id,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,arrival_date,stays_in_weekend_nights,...,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests
158,HB-101806,City Hotel,0,0,2024,November,47,19,2024-11-19,0,...,A,0,No Deposit,9.0,NaN,0,Transient,0.00,0,1
1695,HB-100564,City Hotel,0,88,2024,October,41,7,2024-10-07,6,...,K,7,No Deposit,193.0,NaN,0,Transient-Party,0.00,0,1
2244,HB-092815,City Hotel,0,108,2024,July,27,5,2024-07-05,2,...,K,0,No Deposit,89.0,NaN,0,Transient,77.53,0,0
3953,HB-111028,City Hotel,0,0,2025,May,19,5,2025-05-05,0,...,C,0,No Deposit,NaN,NaN,0,Transient,0.00,0,0
4435,HB-047181,City Hotel,0,0,2024,February,7,12,2024-02-12,0,...,A,2,No Deposit,NaN,45.0,0,Transient,0.00,0,2
4473,HB-111043,City Hotel,0,17,2025,April,17,27,2025-04-27,2,...,K,7,No Deposit,290.0,NaN,0,Transient,0.00,0,0
5078,HB-082265,City Hotel,0,0,2023,December,51,24,2023-12-24,0,...,K,1,No Deposit,28.0,NaN,0,Transient,0.00,0,0
6602,HB-055810,City Hotel,0,256,2024,August,34,21,2024-08-21,3,...,K,1,No Deposit,11.0,NaN,0,Transient,57.30,0,0
8286,HB-107027,City Hotel,0,37,2025,February,8,22,2025-02-22,3,...,K,4,No Deposit,28.0,NaN,0,Transient,0.00,0,0
8659,HB-031766,Resort Hotel,0,31,2024,December,52,27,2024-12-27,2,...,D,3,No Deposit,NaN,NaN,0,Transient,28.00,1,0


hay 37 reservas sin huespedes

In [13]:
if len(sin_huespedes) > 0:
    registrar(
        f"Existencia de {len(sin_huespedes)} reservas sin ningún huésped registrado.",
        "adults, children, babies",
        "Eliminar filas afectadas del dataset de trabajo.",
        "Una reserva requiere al menos 1 huésped registrado para ser comercialmente válida."
    )
    df_limpio = df_limpio.drop(sin_huespedes.index)


# 2. Diagnóstico y Limpieza
## C. supresion de nulos, conversion


In [15]:
# C. supresión de Nulos
# reemplaza valores faltantes y cambia tipo de datos
df_limpio['agent'] = df_limpio['agent'].fillna(0).astype(int)
registrar("Faltantes en Agente", "agent", "Imputar con 0", "Valor nulo indica venta directa sin intermediación de agencia.")

df_limpio['company'] = df_limpio['company'].fillna(0).astype(int)
registrar("Faltantes en Empresa", "company", "Imputar con 0", "Valor nulo indica cliente particular no corporativo.")

df_limpio['country'] = df_limpio['country'].fillna("Unknown")
registrar("Faltantes en País", "country", "Imputar con 'Unknown'", "Evita sesgar la distribución de origen imputando la moda.")

df_limpio['children'] = df_limpio['children'].fillna(0).astype(int)
registrar("Faltantes en Niños", "children", "Imputar con 0", "Se asume ausencia de menores ante la falta de registro explícito.")

# diagnostico y limpieza
## Detección de Tarifas <= 0 y Outliers de Adultos

In [17]:
adr_invalidos = df_limpio[df_limpio['adr'] <= 0]
adults_outliers = df_limpio[df_limpio['adults'] > 10]

In [18]:
print(len(adr_invalidos), len(adults_outliers))

415 2


In [19]:
registrar(
    f"Presencia de {len(adr_invalidos)} registros con ADR menor o igual a 0.",
    "adr",
    "Identificar valores no operativos para aislar en análisis tarifarios.",
    "Un valor de tarifa <= 0 distorsiona el cálculo del precio promedio real."
)
if len(adults_outliers) > 0:
    registrar(
        f"Valores atípicos en adultos (máximo detectado: {df_limpio['adults'].max()}).",
        "adults",
        "Registrar inconsistencia de carga conservando el registro.",
        "Capacidad excede la ocupación estándar; requiere análisis separado."
    )


# 3. Transformación de Datos y Variables Derivadas

In [20]:
df_limpio['arrival_date_parsed'] = pd.to_datetime(df_limpio['arrival_date'], errors='coerce')
df_limpio['total_guests'] = df_limpio['adults'] + df_limpio['children'] + df_limpio['babies']
df_limpio['total_stay_nights'] = df_limpio['stays_in_weekend_nights'] + df_limpio['stays_in_week_nights']
df_limpio['is_family'] = ((df_limpio['children'] + df_limpio['babies']) > 0).astype(int)
df_limpio['estimated_total_cost'] = df_limpio['total_stay_nights'] * df_limpio['adr']

df_limpio['lead_time_group'] = pd.cut(
    df_limpio['lead_time'],
    bins=[-1, 7, 30, 90, 180, 999],
    labels=['Mismo día/Semana', 'Último mes', '1-3 Meses', '3-6 Meses', 'Más de 6 Meses']
)


In [21]:
len(df_limpio)

24963

# 4. Exposición de Bitácora y Estado Final

In [22]:
# 4. Exposición de Bitácora y Estado Final
df_bitacora = pd.DataFrame(registros_bitacora)
display(df_bitacora)
print(f"Dimensiones finales del dataset preparado: {df_limpio.shape[0]} filas y {df_limpio.shape[1]} columnas.")

,problema_detectado,variable_afectada,decision_tomada,justificacion
0,Se detectaron 7332 registros idénticos por com...,Atributos combinados de reserva,Conservar registros registrando el hallazgo.,Pueden responder a reservas simultáneas de gru...
1,Se detectaron 7332 registros idénticos por com...,Atributos combinados de reserva,Conservar registros registrando el hallazgo.,Pueden responder a reservas simultáneas de gru...
2,Existencia de 37 reservas sin ningún huésped r...,"adults, children, babies",Eliminar filas afectadas del dataset de trabajo.,Una reserva requiere al menos 1 huésped regist...
3,Faltantes en Agente,agent,Imputar con 0,Valor nulo indica venta directa sin intermedia...
4,Faltantes en Empresa,company,Imputar con 0,Valor nulo indica cliente particular no corpor...
5,Faltantes en País,country,Imputar con 'Unknown',Evita sesgar la distribución de origen imputan...
6,Faltantes en Niños,children,Imputar con 0,Se asume ausencia de menores ante la falta de ...
7,Faltantes en Agente,agent,Imputar con 0,Valor nulo indica venta directa sin intermedia...
8,Faltantes en Empresa,company,Imputar con 0,Valor nulo indica cliente particular no corpor...
9,Faltantes en País,country,Imputar con 'Unknown',Evita sesgar la distribución de origen imputan...


Dimensiones finales del dataset preparado: 24963 filas y 38 columnas.


El dataset quedó completamente depurado y listo para el modelado, pasando a 24,963 filas (tras eliminar 37 reservas inválidas sin huéspedes) y 38 columnas (por la inclusión de las 6 variables derivadas). Se resolvieron los nulos mediante imputaciones de negocio y se mantuvieron los duplicados grupales e inconsistencias de tarifa para no distorsionar el volumen real de ventas.  